### PR TITLE
Update HCP Packer registry image extraction, validation, publishing logic for registry capable plugins

### PR DESCRIPTION
**Pull request by nywilken**
This version of Packer core has been updated with new error handling to fail a build if Packer is unable to decode or publish image metadata to the HCP Packer Registry. These changes use the development version of https://github.com/hashicorp/packer-plugin-sdk/pull/76.

In these changes the `RegistryBuilder` and `RegistryPostProcessor` have been updated to used mapstructure for decoding all Artifact State data that is to be sent to HCP Packer Registry. If decoding fails for whatever reason the build fails and no data is published to HCP Packer Registry - open question is should we call artifact.Destroy() on publishing image errors? 

Lastly, this change renames the packer_registry pkg to just registry. All references to the package are still set as packerregistry  but the imports have been updated. 

This change also vendors the latest versions of Amazon, GoogleCompute and Azure Arm plugins for the the private beta release.

### CODE REVIEW

**Review by azr — APPROVED** [review]
dope !

**Review by sylviamoss — APPROVED** [review]
LGTM! 👍🏼 
As for the `artifact.Destroy()`, I wouldn't destroy it if it fails unless this is explicitly desired. My thought is that it can be a good source for debugging 🤔  What do you think?

**Comment by nywilken** [comment]
> LGTM! 👍🏼
> As for the `artifact.Destroy()`, I wouldn't destroy it if it fails unless this is explicitly desired. My thought is that it can be a good source for debugging 🤔 What do you think?

Yeah, that is a good point. This also may lead the way to adding some sort of import of existing image data to the registry functionality. Once we have a clear way to fail a build if the builder or post-process is not HCP Packer registry capable the need for destroying will not be necessary. I like your idea of not doing anything unless explicitly desired. Thanks!

**Review by SwampDragons — APPROVED** [review]

File: go.mod [hunk 1]
```diff
@@ -39,14 +38,14 @@ require (
 	github.com/hashicorp/packer-plugin-alicloud v1.0.0
 	github.com/hashicorp/packer-plugin-amazon v1.0.1-dev
 	github.com/hashicorp/packer-plugin-ansible v1.0.0
-	github.com/hashicorp/packer-plugin-azure v1.0.2
+	github.com/hashicorp/packer-plugin-azure v1.0.3-0.20210903201247-40e2012a9e82
```
[the remark this answers is not in the document]
Comment (nywilken) [reply go.mod]: Thanks for calling this out. I updated to the latest release of all plugins now so we should be good here.

**Pull request merged**